---
title: "1. Preprocessing with ArchR"
author: "Yang Dong"
date: "2025-12-17"
output: 
  html_document: 
    toc: yes
    toc_float: yes
    code_folding: hide
editor_options: 
  chunk_output_type: console
---
### Setup

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
```

### Load packages
```{r}
library(ArchR)
library(BSgenome.Hsapiens.UCSC.hg19)
addArchRThreads(threads = 8)
addArchRGenome("hg19")
```

### Check mapped ATAC-seq fragment number
```{r}
nFrag=read.table(here::here("data","scATAC-seq","ALL_BC_PAIR.bam.bedpe.changed.bed.cell_stat.txt"))[,2]
LW=10**4.5
UP=10**5.5
length(which(nFrag>LW & nFrag<UP))
# 516 cells
```

### Create Arrow files
Only the cells with mapped ATAC-seq fragment number between `10**4.5` and `10**5` are maintained.
```{r}
inputFiles=c(here::here("data","scATAC-seq","ALL_BC_PAIR.bam.bedpe.changed.bed.agg.sorted.bed.gz"))
names(inputFiles)='scATAC-seq'
ArrowFiles <- createArrowFiles(
  inputFiles = inputFiles,
  sampleNames = names(inputFiles),
  minTSS = 4, #Dont set this too high because you can always increase later
  minFrags = LW, 
  maxFrags = UP,
  minFragSize = 10,
  maxFragSize = 2000,
  addTileMat = TRUE,
  addGeneScoreMat = TRUE
)
```

### Inferring Doublets
```{r}
doubScores <- addDoubletScores(
  input = ArrowFiles,
  k = 10, #Refers to how many cells near a "pseudo-doublet" to count.
  knnMethod = "UMAP", #Refers to the embedding to use for nearest neighbor search.
  LSIMethod = 1
)
```

### Create an ArchR project
```{r}
proj <- ArchRProject(
  ArrowFiles = ArrowFiles, 
  outputDirectory = "output",
  copyArrows = TRUE #This is recommened so that you maintain an unaltered copy for later usage.
)
```

### Label cells with time (D4 or D8) 
```{r}
tmp=t(matrix(unlist(stringr::str_split(rownames(proj@cellColData),'#')),nrow=2))[,2]
tmp=t(matrix(unlist(stringr::str_split(tmp,'_')),nrow=2))[,1]
time=rep('D4',length(tmp))
time[which(tmp %in% c('P6','P7','P8'))]='D8'
proj@cellColData$time=time
```

### Check the number of reads in TSS 
Only cells with log10Reads in TSS between 3.5 and 4.5 were retained. 
```{r}
Log10ReadsInTSS=log(proj@cellColData$ReadsInTSS,10)
plot(density(Log10ReadsInTSS))
proj <- subsetArchRProject(proj,cells= getCellNames(proj)[which(Log10ReadsInTSS>3.5 & Log10ReadsInTSS <4.5)],force=TRUE)
```

### Remove doublets
2 cells were filtered. 474 cells were retaind. 
```{r}
proj <- filterDoublets(ArchRProj = proj)
```

### Dimensionality Reduction and Clustering
Cells were mainly separated by time. An alternative clustering strategy was desired. 
```{r}
proj <- addIterativeLSI(ArchRProj = proj, useMatrix = "TileMatrix", name = "IterativeLSI")
proj <- addClusters(input = proj, reducedDims = "IterativeLSI")
proj <- addUMAP(ArchRProj = proj, reducedDims = "IterativeLSI",force = T)
plotEmbedding(ArchRProj = proj, colorBy = "cellColData",
              name = "Clusters", embedding = "UMAP",size=4,baseSize=10)
plotEmbedding(ArchRProj = proj, colorBy = "cellColData",
              name = "time", embedding = "UMAP",size=4,baseSize=10)
```

### Add GeneScores
```{r}
proj <- addImputeWeights(proj)

# Plot GeneSCores on UMAP
plotEmbedding(
    ArchRProj = proj, 
    colorBy = "GeneScoreMatrix", 
    name = "GATA6", 
    embedding = "UMAP",
    bins = 60,
    imputeWeights = getImputeWeights(proj)
)  + scale_fill_gradient(low = "white", high = "darkblue")
```

### Call peaks, and Add peak matrix and background peaks
```{r}
proj <- addGroupCoverages(ArchRProj = proj, groupBy = "time",force = TRUE)

pathToMacs2 <- findMacs2()

proj <- addReproduciblePeakSet(
    ArchRProj = proj, 
    groupBy = "time", 
    pathToMacs2 = pathToMacs2,
    force = TRUE
    )
proj <- addPeakMatrix(proj,force = TRUE)
proj <- addBgdPeaks(proj,force = TRUE)
```

### Save ArchR project
```{r}
proj <- saveArchRProject(ArchRProj = proj)
```




