---
title: '5. Use consistency score in zebrafish multi-omics data (Get consistency score)'
author: "Yang"
date: "2025-12-23"
output: 
  html_document: 
    toc: yes
    toc_float: yes
    code_folding: hide
editor_options: 
  chunk_output_type: console
---
### Setup

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
```

### Load packages
```{r}
library(ArchR)
addArchRThreads(threads = 8)
source('https://raw.githubusercontent.com/jumphone/InferLoop/main/inferloop/InferLoop.R')
source('https://gitee.com/jumphone/BEER/raw/master/BEER.R')
library(readr)
library(TFBSTools)
library(Seurat)
library(dplyr)
library(Metrics) 
library(aricode)
library(mclust)
```

### Load data
```{r}
proj <- loadArchRProject(path = here::here("ArchRSubset"))
```

### Get Gene scores
```{r}
GeneScore = getMatrixFromProject(proj,useMatrix ='GeneScoreMatrix')
GeneScore_matrix = GeneScore@assays@data$GeneScoreMatrix %>% as.matrix()
rownames(GeneScore_matrix)= GeneScore@elementMetadata$name
```

### Load zebrafish motif annotation
```{r}
load(here::here("reference","danRer11.Motif2.RData"))
```

### Add motif annotation to ArchRProject using defined PWMatrixList objects
```{r}
proj <- addMotifAnnotations(ArchRProj = proj, motifPWMs = danRer11.Motif2, name = "Motif",force = TRUE)

proj <- addDeviationsMatrix(
      ArchRProj = proj, 
      peakAnnotation = "Motif",
      force = TRUE
      )
```

### Get motif enrichment scores
915 motifs identified
```{r}
Motif = getMatrixFromProject(proj, useMatrix ='MotifMatrix')
Motif_matrix = Motif@assays@data$z
Motif_matrix_TF=rownames(Motif_matrix)
```

### Construct consistency score
```{r}
# Combine GeneScore and motif enrichment score, using common genes for the consistency score
COM=.simple_combine(GeneScore_matrix,Motif_matrix)
D1=COM$exp_sc_mat1
D2=COM$exp_sc_mat2

consistency_score=D1
i=1
while(i<=nrow(consistency_score)){
    this_x=D1[i,]
    this_y=D2[i,]
    this_z=inferloop.calILS(this_x,this_y,only_pos=TRUE)
    consistency_score[i,]=this_z
    i=i+1}

colnames(consistency_score) <- str_remove(colnames(consistency_score), "zebrafish_scATAC-seq#")
consistency_score_zscore=t(apply(consistency_score,1,scale,center=F))
colnames(consistency_score_zscore)=colnames(consistency_score)
```

### Save data
```{r}
# saveRDS(k_clusters, here::here("output","TMP.CLST.rds"))
saveRDS(Motif_matrix,here::here("output","Motif_matrix_zebrafish.rds"))
saveRDS(consistency_score, here::here("output","consistency_score_zebrafish.rds"))
```

### Using consistency score for clustering
```{r}
Embryo_6h <- readRDS(here::here("data","Embryo_6h.rds"))

Embryo_6h_subset <- subset(Embryo_6h, subset = cell_types == "epiblast-MT-high" | cell_types == "ectoderm-ventral-MT-high", invert =T )
# Merge cell types
Embryo_6h_subset$cell_types <- sub("non-dorsal margin involuted","non-dorsal margin",Embryo_6h_subset$cell_types) 
Embryo_6h_subset$cell_types <- sub("ectoderm-ventral","ectoderm",Embryo_6h_subset$cell_types) 
Embryo_6h_subset$cell_types <- sub("ectoderm-dorsal","ectoderm",Embryo_6h_subset$cell_types) 

Embryo_6h_subset$cell_types <- sub("non-dorsal margin","margin",Embryo_6h_subset$cell_types) 
Embryo_6h_subset$cell_types <- sub("dorsal margin-adaxial mesoderm", "margin",Embryo_6h_subset$cell_types) 
Embryo_6h_subset$cell_types <- sub("endoderm", "endo&axis_meso",Embryo_6h_subset$cell_types) 
Embryo_6h_subset$cell_types <- sub("axis mesoderm", "endo&axis_meso",Embryo_6h_subset$cell_types) 


DefaultAssay(Embryo_6h_subset) <- "RNA"
Idents(Embryo_6h_subset) <- "cell_types"
markers <- FindAllMarkers(Embryo_6h_subset, only.pos = TRUE) 
markers %>%
  dplyr::filter(.$gene %in% Motif_matrix_TF) %>%
    group_by(cluster) %>%
#    dplyr::filter(avg_log2FC > 1) %>%
    slice_head(n = 10) %>%
    ungroup() -> top10
DefaultAssay(Embryo_6h_subset) <- "RNA"

DoHeatmap(Embryo_6h_subset, features = top10$gene) + NoLegend()



MARKER =c('grhl1', "klf4","nr1i2", "cebpb",        # EVL
         "ebf3a","sox19b", "insm1b",               # ectoderm
         "mespab","mespbb", "tbx16", "snai1a",     # margin
         'foxa',"foxa2","gata5","foxa3"            # endoderm & axis mesoderm

         )

MARKER.TAG=c(rep('EVL',4), rep("ectoderm",3), rep("margin",4), rep("endoderm & axis mesoderm",4))

consistency_score_markers=consistency_score_zscore[MARKER,]
markers_mean=.generate_mean(t(consistency_score_markers),MARKER.TAG)
COR=cor(t(markers_mean))
COR[which(is.na(COR))]=0
# Force clustering into 5 groups
set.seed(1)
k_clusters = cutree(hclust(dist(COR)), k=4) 
```

### Heatmap
```{r}
ht=.drawHeatmap2(consistency_score_markers, column_split=k_clusters,row_split=MARKER.TAG,
                show_column_names=F, show_row_names= T,
                show_column_dend = F, show_row_dend =T,
                cluster_columns=F, cluster_rows=T,
                )

ht.z=.drawHeatmap3(consistency_score_markers, column_split=k_clusters,row_split=MARKER.TAG,
                show_column_names=F, show_row_names= T,
                show_column_dend = F, show_row_dend =T,
                cluster_columns=F, cluster_rows=T,
                )

ht.mean.gene=.splitMean(ht.z, rownames(consistency_score_markers), k_clusters)
print(ht.mean.gene)
```


### Annotate cell types
```{r}
Embryo_6h_subset <- subset(Embryo_6h_subset, cells = colnames(consistency_score))

k_cluster_df <- k_clusters %>% as.data.frame() 
k_cluster_df <-  k_cluster_df[colnames(Embryo_6h_subset),] %>% as.data.frame()



Embryo_6h_subset$k_cluster <- k_cluster_df$.
Idents(Embryo_6h_subset) <- "k_cluster"
DimPlot(Embryo_6h_subset,reduction = "umap", label = F)


EVL.ID=which(k_clusters==1)
ectoderm.ID=which(k_clusters==2)
margin.ID=which(k_clusters==4)
endo_axis_meso.ID=which(k_clusters==3)
CLST=rep('NA',length(k_clusters))
CLST[EVL.ID]='EVL'
CLST[ectoderm.ID]='ectoderm'
CLST[margin.ID]='margin'
CLST[endo_axis_meso.ID]='endo&axis_meso'


k_cluster_df_anno <- k_clusters %>% as.data.frame() 
k_cluster_df_anno$anno <- CLST
k_cluster_df_anno <-  k_cluster_df_anno[colnames(Embryo_6h_subset),] %>% as.data.frame()

# UMAP of consistency score
Embryo_6h_subset$anno <- k_cluster_df_anno$anno
Idents(Embryo_6h_subset) <- "anno"
DimPlot(Embryo_6h_subset,reduction = "umap", label = F) + NoLegend() 


### UMAP of RNA
Idents(Embryo_6h_subset) <- "cell_types"
DimPlot(Embryo_6h_subset,reduction = "umap", label = F) + NoLegend() 
```

### Access the clustering performance using ARI
```{r}
meta <- Embryo_6h_subset@meta.data

adjustedRandIndex(meta$cell_types, meta$anno) 
# 0.39
```

# Showing consistency score on UMAP
```{r}
consistency_score_zscore_fileter <- consistency_score_zscore[,colnames(Embryo_6h_subset)]

Embryo_6h_subset[["consistency_score"]] <- CreateAssayObject(data = consistency_score_zscore_fileter)

DefaultAssay(Embryo_6h_subset) <- "consistency_score"
FeaturePlot(Embryo_6h_subset, features = "nr1i2",cols = c("lightgrey", "#a70000"))

FeaturePlot(Embryo_6h_subset, features = "tbx16",cols = c("lightgrey", "darkgreen"))

FeaturePlot(Embryo_6h_subset, features = "ebf3a", cols = c("lightgrey", "#0080fe"))

FeaturePlot(Embryo_6h_subset, features = "foxa",cols = c("lightgrey", "#6F2DA8"))
```

### Clustering using gene activity 
```{r}
colnames(GeneScore_matrix) <- str_remove(colnames(GeneScore_matrix), "zebrafish_scATAC-seq#")
GeneScore_matrix_z_score=t(apply(GeneScore_matrix,1,scale,center=F))
colnames(GeneScore_matrix_z_score)=colnames(GeneScore_matrix)

GeneScore_matrix_z_score_markers =GeneScore_matrix_z_score[MARKER,]
markers_mean=.generate_mean(t(GeneScore_matrix_z_score_markers),MARKER.TAG)
COR=cor(t(markers_mean))
COR[which(is.na(COR))]=0
# Force clustering into 5 groups
set.seed(1)
k_clusters_gene_activity = cutree(hclust(dist(COR)), k=4) 

k_cluster_gene_activity_df <- k_clusters_gene_activity %>% as.data.frame() 
k_cluster_gene_activity_df <- k_cluster_gene_activity_df[colnames(Embryo_6h_subset),] %>% as.data.frame()

Embryo_6h_subset$k_cluster_gene_activity <- k_cluster_gene_activity_df$.
Idents(Embryo_6h_subset) <- "k_cluster_gene_activity"
DimPlot(Embryo_6h_subset,reduction = "umap", label = F)


EVL.ID=which(k_clusters_gene_activity==1)
ectoderm.ID=which(k_clusters_gene_activity==2)
margin.ID=which(k_clusters_gene_activity==3)
endo_axis_meso.ID=which(k_clusters_gene_activity==4)
CLST=rep('NA',length(k_clusters_gene_activity))
CLST[EVL.ID]='EVL'
CLST[ectoderm.ID]='ectoderm'
CLST[margin.ID]='margin'
CLST[endo_axis_meso.ID]='ectoderm'


k_clusters_gene_activity_df_anno <- k_clusters_gene_activity %>% as.data.frame() 
k_clusters_gene_activity_df_anno$anno <- CLST
k_clusters_gene_activity_df_anno <-  k_clusters_gene_activity_df_anno[colnames(Embryo_6h_subset),] %>% as.data.frame()

# UMAP of gene activity
Embryo_6h_subset$anno_gene_activity <- k_clusters_gene_activity_df_anno$anno
Idents(Embryo_6h_subset) <- "anno_gene_activity"
DimPlot(Embryo_6h_subset,reduction = "umap", label = F) + NoLegend() 

# Calculating ARI
meta <- Embryo_6h_subset@meta.data
adjustedRandIndex(meta$cell_types, meta$anno_gene_activity) 
```

### Clustering using binding activity
```{r}
colnames(Motif_matrix) <- str_remove(colnames(Motif_matrix), "zebrafish_scATAC-seq#")
Motif_matrix <- Motif_matrix %>% as.data.frame()
Motif_matrix_z_score=t(apply(Motif_matrix,1,scale,center=F))
colnames(Motif_matrix_z_score)=colnames(Motif_matrix)

Motif_matrix_z_score_markers =Motif_matrix_z_score[MARKER,]
markers_mean=.generate_mean(t(Motif_matrix_z_score_markers),MARKER.TAG)
COR=cor(t(markers_mean))
COR[which(is.na(COR))]=0
# Force clustering into 5 groups
set.seed(1)
k_clusters_binding_activity = cutree(hclust(dist(COR)), k=4) 


k_cluster_binding_activity_df <- k_clusters_binding_activity %>% as.data.frame() 
k_cluster_binding_activity_df <- k_cluster_binding_activity_df[colnames(Embryo_6h_subset),] %>% as.data.frame()

Embryo_6h_subset$k_cluster_binding_activity <- k_cluster_binding_activity_df$.
Idents(Embryo_6h_subset) <- "k_cluster_binding_activity"
DimPlot(Embryo_6h_subset,reduction = "umap", label = F)


EVL.ID=which(k_clusters_binding_activity==1)
ectoderm.ID=which(k_clusters_binding_activity==2)
margin.ID=which(k_clusters_binding_activity==3)
endo_axis_meso.ID=which(k_clusters_binding_activity==4)
CLST=rep('NA',length(k_clusters_binding_activity))
CLST[EVL.ID]='EVL'
CLST[ectoderm.ID]='ectoderm'
CLST[margin.ID]='margin'
CLST[endo_axis_meso.ID]='endo&axis_meso'


k_clusters_binding_activity_df_anno <- k_clusters_binding_activity %>% as.data.frame() 
k_clusters_binding_activity_df_anno$anno <- CLST
k_clusters_binding_activity_df_anno <-  k_clusters_binding_activity_df_anno[colnames(Embryo_6h_subset),] %>% as.data.frame()

# UMAP of binding activity
Embryo_6h_subset$anno_binding_activity <- k_clusters_binding_activity_df_anno$anno
Idents(Embryo_6h_subset) <- "anno_binding_activity"
DimPlot(Embryo_6h_subset,reduction = "umap", label = F) + NoLegend() 

# Calculating ARI
meta <- Embryo_6h_subset@meta.data
adjustedRandIndex(meta$cell_types, meta$k_cluster_binding_activity) 
```

### Pie chart showing proportions of each cell type
```{r}
df <- prop.table(table(Embryo_6h_subset@meta.data$cell_types)) %>% as.data.frame()
df$Var1 <- factor(df$Var1, c("EVL","margin","ectoderm","endo&axis_meso"))
df <- prop.table(table(Embryo_6h_subset@meta.data$anno)) %>% as.data.frame()
df$Var1 <- factor(df$Var1, c("EVL","margin","ectoderm","endo&axis_meso"))
df <- prop.table(table(Embryo_6h_subset@meta.data$anno_gene_activity)) %>% as.data.frame()
df$Var1 <- factor(df$Var1, c("EVL","margin","ectoderm","endo&axis_meso")) 
df <- prop.table(table(Embryo_6h_subset@meta.data$anno_binding_activity)) %>% as.data.frame()
df$Var1 <- factor(df$Var1, c("EVL","margin","ectoderm","endo&axis_meso"))
df$labels <- scales::percent(df$Freq)

ggplot(df, aes(x = "", y = Freq, fill = Var1)) +
  geom_col() +
  geom_text(aes(label = labels),
            position = position_stack(vjust = 0.5)) +
  coord_polar(theta = "y") + 
  theme_void() 
```

### Bar plot show the percentage of matched cells in each cell type
```{r}
EVL_pro <- sum(Embryo_6h_subset@meta.data$cell_types == "EVL" & Embryo_6h_subset@meta.data$anno_binding_activity == "EVL")/sum(Embryo_6h_subset@meta.data$cell_types == "EVL")
margin_pro <- sum(Embryo_6h_subset@meta.data$cell_types == "margin" & Embryo_6h_subset@meta.data$anno_binding_activity == "margin")/sum(Embryo_6h_subset@meta.data$cell_types == "margin")
ecto_pro <- sum(Embryo_6h_subset@meta.data$cell_types == "ectoderm" & Embryo_6h_subset@meta.data$anno_binding_activity == "ectoderm")/sum(Embryo_6h_subset@meta.data$cell_types == "ectoderm")
endo_pro <- sum(Embryo_6h_subset@meta.data$cell_types == "endo&axis_meso" & Embryo_6h_subset@meta.data$anno_binding_activity == "endo&axis_meso")/sum(Embryo_6h_subset@meta.data$cell_types == "endo&axis_meso")

df <- data.frame(x = c("EVL","margin","ectoderm","endo&axis_meso"),
                 y = c(EVL_pro,margin_pro,ecto_pro,endo_pro))
df$x <- factor(df$x, c("EVL","margin","ectoderm","endo&axis_meso"))
df$labels <- scales::percent(df$y)

ggplot(df, aes(x = x, y = y, fill = x)) +
  geom_bar(stat = "identity",width=0.5) + 
  ylim(c(0,1)) +
  geom_text(aes(label = labels)) +
  theme_bw() +
  theme(axis.line = element_line(colour = "black"),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_blank(),
    panel.background = element_blank()) +
  NoLegend()

```


### Functions used to draw heatmap
```{r}
.drawHeatmap2<-function(o.mat, ...){
    library('ComplexHeatmap')
    library('circlize')
    o.mat=o.mat
    #library('dbscan')
    ###################
    vvv=as.numeric(o.mat)
    col_fun =colorRamp2(
            c(-1,-0.5,0,0.5,1),
               c('royalblue3','skyblue1',
              'white','indianred1','red3'))
    ################
    ht=Heatmap(o.mat,row_title='',name="v",
        #cluster_rows=T, cluster_columns=T,
        #show_column_dend = T, show_row_dend = T,
        #show_column_names= T, show_row_names= F,
        #row_split=R.SPLIT,
        #column_split=C.SPLIT,
        col=col_fun, border = TRUE,
        row_names_side='right',
        row_names_gp = gpar(fontsize = 10, lineheight=NULL),
        column_names_gp = gpar(fontsize = 10, lineheight=NULL),
        row_names_max_width = max_text_width(rownames(o.mat),gp = gpar(fontsize = 15)),
        column_names_max_height = max_text_width(colnames(o.mat),gp = gpar(fontsize = 15)),
        ...)
    pdf('tmp_initialize.pdf',width=7,height=7)
    ht=draw(ht)
    dev.off()
    return(ht)
}

.drawHeatmap3<-function(o.mat, ...){
    library('ComplexHeatmap')
    library('circlize')
    o.mat=o.mat
    #library('dbscan')
    ###################
    vvv=as.numeric(o.mat)
    col_fun =colorRamp2(
            c(qnorm(0.5),qnorm(0.6),qnorm(0.7),qnorm(0.8),qnorm(0.9)),
               c('royalblue3','skyblue1',
              'white','indianred1','red3'))
    ################
    ht=Heatmap(o.mat,row_title='',name="v",
        #cluster_rows=T, cluster_columns=T,
        #show_column_dend = T, show_row_dend = T,
        #show_column_names= T, show_row_names= F,
        #row_split=R.SPLIT,
        #column_split=C.SPLIT,
        col=col_fun, border = TRUE,
        row_names_side='right',
        row_names_gp = gpar(fontsize = 10, lineheight=NULL),
        column_names_gp = gpar(fontsize = 10, lineheight=NULL),
        row_names_max_width = max_text_width(rownames(o.mat),gp = gpar(fontsize = 15)),
        column_names_max_height = max_text_width(colnames(o.mat),gp = gpar(fontsize = 15)),
        ...)
    pdf('tmp_initialize.pdf',width=7,height=7)
    ht=draw(ht)
    dev.off()
    return(ht)
}

.splitMean <-function(ht, R.SPLIT, C.SPLIT){
    UR.SPLIT=sort(unique(R.SPLIT))
    UC.SPLIT=sort(unique(C.SPLIT))
    ht_mat=ht@ht_list$v@matrix
    i=1
    while(i<=length(UR.SPLIT)){
        j=1
        while(j<=length(UC.SPLIT)){
            this_r=UR.SPLIT[i]
            this_c=UC.SPLIT[j]
            this_r_index=which(R.SPLIT==this_r)
            this_c_index=which(C.SPLIT==this_c)
            this_r_name=rownames(ht_mat)[this_r_index]
            this_c_name=colnames(ht_mat)[this_c_index]
            this_mean=mean(as.numeric(ht_mat[this_r_index,this_c_index]))
            ht_mat[this_r_name,this_c_name]=this_mean
        j=j+1}
    i=i+1}
    ht@ht_list$v@matrix=ht_mat
    return(ht)
}
```